compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -578,6 +579,43 @@ declare_lint! {
     crate_level_only
 }
 
+declare_lint! {
+    /// The `incompatible_target_modifiers` lint detects crates with incompatible target modifiers
+    /// (abi-changing or vulnerability-affecting flags).
+    ///
+    /// ### Example
+    ///
+    /// ```rust,ignore (needs extern crate)
+    /// #![deny(incompatible_target_modifiers)]
+    /// ```
+    ///
+    /// When main and dependency crates are compiled with `-Zregparm=1` and `-Zregparm=2` correspondingly.
+    ///
+    /// This will produce:
+    ///
+    /// ```text
+    /// error: crate `incompatible_regparm` has incompatible target modifier with extern crate `wrong_regparm`: `regparm = ( Some(1) | Some(2) )`
+    /// --> $DIR/incompatible_regparm.rs:5:1
+    ///   |
+    /// 1 | #![no_core]
+    ///   | ^
+    ///   |
+    ///   = note: `#[deny(incompatible_target_modifiers)]` on by default
+    /// ```
+    ///
+    /// ### Explanation
+    ///
+    /// `Target modifiers` are compilation flags that affects abi or vulnerability resistance.
+    /// Linking together crates with incompatible target modifiers would produce incorrect code
+    /// or degradation of vulnerability resistance.
+    /// So this lint should find such inconsistency.
+    ///
+    pub INCOMPATIBLE_TARGET_MODIFIERS,
```

Comment:
Why is this a lint?

Reply:
Well, I just used `lint_unused_crate_dependency` as a reference of between-crates problem. Is it a problem to be lint? How should it be implemented in a better way?

Reply:
I'd like to solicit some other opinions about this so asked in the RFC - https://github.com/rust-lang/rfcs/pull/3716/files#r1866108577

Reply:
I'd like this to be changed to be a hard error rather than a lint